source/iotagent-architecture.rst

```diff
@@ -0,0 +1,543 @@
+trusted by dojot. Once this certificate is valid, device identity can be
+checked in two forms:
+
+- Device ID encoded in certificate: although this is a less-reliable mechanism,
+  it allows greater flexibility using many devices in a controlled deployment.
+  This is based on setting the common name (CN certificate field) as dojot
+  device ID. Therefore, IoT agent should check whether this device exists or
+  not and allow or deny the connection right away depending on this
+  verification. The weak points of this mechanisms is that the device
+  certificate must be signed by dojot's internal CA (once there is a procedure
+  to sign only one certificate per device) and, if this certificate is valid,
+  then its ID must also be valid. If any other CA is used, then this mechanism
+  has no valid use.
+
+- IoT agent has all valid certificates: if an administrator wants to use an
+  external CA to sign all device certificates, then there is no actual control
+  of which device ID was used to generate a particular certificate. Therefore,
+  IoT agent must have all valid certificates properly mapped onto a device list
+  - this will guarantee that only one certificate is allowed to a particular
+  device and vice-versa.
+
+Using the first mechanism, the device (or an operator configuring a device for
+the first time) must call dojot CA to generate a signed certificate for itself.
+There is no further action for IoT agent to take as long as dojot CA is used.
+
+The second mechanism, however, requires that an IoT agent offer methods to
+manage certificates. The developer must take into account also that this IoT
+agent must be able to scale - these certificates must be accessible to all IoT
+agent instances, if allowed by deployment.
+
+Communication security
+----------------------
+
+With a valid certificate, a device can create a communication channel with
+dojot. For connection-oriented channels, this certificate should be used
+alongside cryptographic keys in order to provide an encrypted channel. For
+other channel types (such exchanging messages through a device gateway, such as
+LoRa or sigfox), it suffice to be sure that the connection between dojot and
+the backend server is secure. The backend identity should be asserted
+beforehand. Once it is known to be trusted, all its messages can be processed
+with no major concern.
+
+Certificate revocation
+----------------------
+
+An IoT agent should be able to be informed about revoked certificates. It
+should expose an API or configuration messages to allow such thing. It should
+not allow any communication with a particular device that uses a revoked
+certificate.
+
+Information context separation
+==============================
+
+A tenant could be thought simply as a group of users that share some resources.
+But its meaning goes beyond that - it implies that these resources do not share
+any common infrastructure (considering anything that transmits, processes or
```

Comment:
it shares an iot agent and the entrypoint ip address

Reply:
But it might not. One would say that is necessary to allocate a new IoT agent for each tenant. I'll add some 'may' and 'might' to alleviate this.